sklearn/datasets/value_dropper.py

```diff
@@ -0,0 +1,243 @@
+# Author : Raghav RV <rvraghav93@gmail.com>
+#
+# Licence : BSD 3 clause
+
+import numpy as np
+
+from ..utils import check_X_y
+from ..utils import check_random_state
+from ..utils.multiclass import type_of_target
+
+
+__all__ = ["drop_values"]
+
+
+def drop_values(X, y=None,
+                missing_values="NaN",
+                missing_rate=0.1,
+                class_correlation=1.0,
+                correlated_classes=1,
+                return_correlated_classes=False,
+                copy=False,
+                random_state=None):
+    """Drop values based on a preset strategy.
+
+    Attributes
+    ----------
+
+    X : ndarray like of shape (n_features, n_samples)
+        Data, in which the values must be dropped and set to
+        ``missing_values``.
+
+    y : ndarray like of shape (n_samples,), optional
+        Target, in a supervised classification task.
+
+        It can also be used for regression targets if the ``class_correlation``
+        is 0. (MCAR missingness).
+
+    missing_values : {"NaN" (or np.nan) | int}, default "NaN"
+        The value to insert to indicate missingness.
+
+    missing_rate : float in range [0, 1], default 0.1
+        The absolute fraction of values that must be missing.
+
+        If there are previously existing missing values (say x fraction of
+        values were already missing), an additional ``missing_rate - x``
+        fraction of values will be dropped further to achieve a total of
+        ``missing_rate`` fraction of missing values.
+
+        That is, at the end, the dataset will contain a total of
+        ``(X.shape[0] * X.shape[1]) * missing_rate`` numbers of missing values.
+
+    class_correlation : float, default 1.0
+        1 (MNAR) - Randomly choose ``n_classes`` (or take from given
+        classes) and correlate the missing values with the occurence
+        of those class classes.
+
+        0 (MCAR) - Randomly drop values without correlating the
+        missingness with any class classes.
+
+        Any value inbetween would constitute a noisy MNAR missingness.
```

Comment:
in terms of API I would parametrize the prob of having a missing for a class using a https://en.wikipedia.org/wiki/Multinomial_distribution

it could be as for class_weight a dict : class_k -> prob_k

API would be simpler:

missing_rate + missing_class_proba

and I don't see the need to return anything else than X and y.